### PR TITLE
Feature: Stateful factory guard

### PR DESCRIPTION
**Pull request by thehobojoe**
### Work done
- Remove factory guard widget
- Move all functionality to a gadget, and add it as a command
- Default off, non-configurable
- Togglable by hotkey (ctrl+g on grid) or clicking in the order menu

#### Test steps
- [ ] Produce cons from factory with factoryguard off, expect them to move to rally point
- [ ] Produce cons from factory with factoryguard on, expect them to automatically guard
- [ ] Interfere with queued factory guard commands and ensure con takes orders as expected
- [ ] Ensure factoryguard command does not show up on non-factory units
- [ ] Ensure factoryguard command does not show up on factories without builders (eg. t3 labs)

image

### Notes
I plan to use this pattern for all custom commands that currently use a renamed on-off command to fix accidental toggle bugs, e.g. with hounds and radar bots, so I would like feedback on this approach. It is mostly stolen from ZK so it should be okay, but altered the code when porting it to BAR so there may be things I missed.

### CODE REVIEW

**Comment by StanczakDominik** [comment]
- [x] Produce cons from factory with factoryguard off, expect them to move to rally point
- [x] Produce cons from factory with factoryguard on, expect them to automatically guard
- [x] Interfere with queued factory guard commands and ensure con takes orders as expected
- [x] Ensure factoryguard command does not show up on non-factory units
- [x] Ensure factoryguard command does not show up on factories without builders (eg. t3 labs)

**Comment by StanczakDominik** [comment]
I think it's lovely! My only concern is this:

> Default off, non-configurable

I know you don't like it yourself, but I think it's a very good "training wheel" for new players, and I think making that default (which I've put a bunch of effort into convincing people about... 😉) has been overall a net positive. Would it be a huge PITA to keep it default on and make it configurable, just like there's "Gremlin default cloak on" under game settings?

File: luaui/Widgets/gui_options.lua [hunk 1]
```diff
@@ -4275,11 +4275,9 @@ function init()
 			end,
 		},
 
-		{ id = "factoryguard", group = "game", category = types.basic, widget = "FactoryGuard", name = Spring.I18N('ui.settings.option.factory') .. widgetOptionColor .. "  " .. Spring.I18N('ui.settings.option.factoryguard'), type = "bool", value = GetWidgetToggleValue("FactoryGuard"), description = Spring.I18N('ui.settings.option.factoryguard_descr') },
```
Comment (StanczakDominik) [inline luaui/Widgets/gui_options.lua:4278]: Keep plz? :(

File: luarules/gadgets/unit_factory_guard.lua [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	GuardFactory(unitID, unitDefID, factID, factDefID)
+end
+
+
+--------------------------------------------------------------------------------
+-- Unit Handling
+
+function gadget:UnitCreated(unitID, unitDefID, _)
+	if isFactory[unitDefID] then
+		factoryGuardCmdDesc.params[1] = 0
```
Comment (StanczakDominik) [inline luarules/gadgets/unit_factory_guard.lua:171]: Is this where that settings-sourced default would be applied?

**Comment by ZephyrSkies7** [comment]
- [x] Produce cons from factory with factoryguard off, expect them to move to rally point
- [x] Produce cons from factory with factoryguard on, expect them to automatically guard
- [x] Interfere with queued factory guard commands and ensure con takes orders as expected
- [x] Ensure factoryguard command does not show up on non-factory units
- [x] Ensure factoryguard command does not show up on factories without builders (eg. t3 labs)

**Comment by ZephyrSkies7** [comment]
Overall, lovely. Even rapidly switching between factory guard on and off doesn't seem to faze it or cause it to behave in unexpected ways. I agree with [StanczakDominik](https://github.com/StanczakDominik) that it should default to on, or perhaps have a setting where we set it such that it is default on or off?

**Comment by thehobojoe** [comment]
I figured the main notes would be the configurability. The main reason I took it away is that gadgets aren't configurable in that way, and I was trying to do the work in a gadget to bypass a couple other problems.

My reasoning is basically: 
- BAR is filled to the brim with nearly-useless settings, we should be removing them at any reasonable opportunity
- When building your first factory, you will have essentially nothing to do for a good 15 seconds. Clicking a single toggle on the lab is not a lot to ask at this stage
- When building later factories, you most likely want this off, because the conditions that make it useful are mostly limited to the first few minutes of a match.

If you feel strongly enough about it I can bring it back, but it will be.. not very nice. A messy widget/gadget pair that will make it a lot easier to sneak bugs in.

**Review by StanczakDominik — APPROVED** [review]
Well, if it calls for nasty workarounds, let's not go there and get it in instead - I would just like to make sure that this is well communicated.

- [ ] microblog
- [ ] discord announcements
- [ ] maybe chobby news post?

It'll give people a nasty surprise the first few times :P

**Pull request merged**